ARROW-7495: [Java] Remove "empty" concept from ArrowBuf, replace with custom referencemanager

tianchen92 opened the pull request:
Related to [ARROW-7495](https://issues.apache.org/jira/browse/ARROW-7495).

With the introduction of ReferenceManager in the codebase, the need for a separate ArrowBuf is no longer necessary. Instead, once can create a new reference manager that is used for the empty ArrowBuf. For reminder/review, empty arrowbufs have a special behavior in that they don't actually have any reference counting semantics and always stay at one. This allow us to better troubleshoot unallocated memory than what would otherwise be an NPE after calling ValueVector.clear()

tianchen92 commented:
cc @jacques-n

emkornfield commented:
Ping @jacques-n do you think you would have time to review this?  @rymurr are you familiar with this part of the code and can do a first pass?

rymurr commented:
Hey @emkornfield. I'm happy to take a look. I likely won't get to it till Monday as I am traveling at the moment.

siddharthteotia commented:
Can we please add a test that tells how this will look like? I mean the proposal is to remove the isEmpty flag from ArrowBuf and have the users (whoever wants an empty ArrowBuf) pass a custom reference manager that is as good as an ArrowBuf.

It will be great if we have some test to do this. I am not sure if a new concrete implementation of ReferenceManager should be written to manage empty bufs or can we use the existing ReferenceManager.NO_OP?

tianchen92 commented:
> Can we please add a test that tells how this will look like?
> 
> IIUC, the proposal is to remove the isEmpty flag from ArrowBuf and have the users (whoever wants an empty ArrowBuf) pass a custom reference manager that is as good as an empty ArrowBuf.
> 
> It will be great if we have some test to do this. I am not sure if a new concrete implementation of ReferenceManager should be written to manage empty bufs or can we use the existing ReferenceManager.NO_OP?

Thanks, IMO, one could either use ReferenceManager.NO_OP or write their own ReferenceManager for empty ArrowBuf. I added a simple test case using ReferenceManager.NO_OP.

rymurr commented:
This looks good to me. Potential nice to have: a static method to generate a default empty buffer of length `n`

tianchen92 commented:
> This looks good to me. Potential nice to have: a static method to generate a default empty buffer of length `n`

Thanks, done.

liyafan82 commented on java/memory/src/main/java/io/netty/buffer/ArrowBuf.java:
```diff
@@ -86,17 +85,14 @@
       final long length,
-      final long memoryAddress,
-      boolean isEmpty) {
+      final long memoryAddress) {
```
I am not sure if directly changing a public API would cause client code break.

tianchen92 replied:
Agree we can't easily change the public api, but i think it's ok if it‘s really needed, and users could correct the behavior when they bump version.

siddharthteotia commented:
@tianchen92 , can you please revise thia? Looks like the work is completed

tianchen92 commented:
> @tianchen92 , can you please revise thia? Looks like the work is completed

ok, revised, thanks!

jacques-n commented:
I think it would be good to have a set of tests ensuring that the behavior of the empty buffer is consistent. For example, refcnt, release/allocate, etc. That way, if someone changes the noop reference manager which this is depending on in the future, we don't break the expected behavior of an EmptyBuf as used by all the vector classes. Other than that, this looks good.

tianchen92 commented:
> I think it would be good to have a set of tests ensuring that the behavior of the empty buffer is consistent. For example, refcnt, release/allocate, etc. That way, if someone changes the noop reference manager which this is depending on in the future, we don't break the expected behavior of an EmptyBuf as used by all the vector classes. Other than that, this looks good.

Good suggestion, seems empty ArrowBuf was used in vectors before populating data. I have added a set of tests in TestValueVector.

emkornfield commented:
@jacques-n are you happy with the test coverage?  Any other issues?

jacques-n commented:
> Good suggestion, seems empty ArrowBuf was used in vectors before populating data. I have added a set of tests in TestValueVector.

The tests you added are all vector based. I'd like to tests against the "empty" ArrowBuf as well.

tianchen92 commented:
> > Good suggestion, seems empty ArrowBuf was used in vectors before populating data. I have added a set of tests in TestValueVector.
> 
> The tests you added are all vector based. I'd like to tests against the "empty" ArrowBuf as well.

I see, I added more cases in TestArrowBuf

jacques-n commented:
Looks mostly good. +1

As a follow-on, it'd be nice to include a test which ensures that the Netty Allocator returns an empty-behaving byte buffer when users allocate a zero byte buffer.

tianchen92 commented:
> include a test which ensures that the Netty Allocator returns an empty-behaving byte buffer when users allocate a zero byte buffer

Thanks jacques, I created https://issues.apache.org/jira/browse/ARROW-8870 to track.
@emkornfield Please help merge this PR when you have time, thanks a lot!

siddharthteotia commented:
@tianchen92 can you rebase this once? I will merge then.

tianchen92 commented:
> @tianchen92 can you rebase this once? I will merge then.

Thanks, rebased:)
